src/Salsa.jl

```diff
@@ -191,16 +207,23 @@ function current_trace(db::Runtime)
 end
 
 """
-    trace_with_key(f::Function, db::Runtime, dbkey)
+    trace_with_key(f::Function, rt::Runtime, dbkey)
 Call `f()` surrounded by calls to `push_key` and `pop_key`. If any exceptions are raised
 pop the unused stack entry and then rethrow the error.
 """
-function trace_with_key(f::Function, db::Runtime, dbkey)
-    push_key(db,dbkey)
+function trace_with_key(f::Function, rt::Runtime, dbkey)
+    push_key(rt, dbkey)
     try
-       f()
+        f()
+    catch e
+        # Wrap the exception in a Salsa exception (at the lowest layer).
+        if !(e isa SalsaDerivedException)
+            rethrow(SalsaDerivedException{typeof(e)}(e, copy(rt.active_query)))
```

Comment:
Ah interesting, I didn't realize you can rethrow a different exception and keep the stacktrace of the wrapped exception. This addresses problems with loosing the stacktrace of the wrapped exception.

Reply:
Yeah, exactly! i've wondered why our exception types contain the stacktrace for a while, but never got around to asking about it. is this why? Maybe we could remove that, then. :)

Reply:
This definitely reduces the need for that, however in other major languages stacktraces are a property of the thrown exception, which means that it would also be available if you leave the catch and start normal control-flow with the exception as a value.